File: Step 04 Binary Search/Step 4.3 Find Answers by BS in Search Space/Find the Smallest Divisor Given a Threshold/Find the Smallest Divisor Given a Threshold.java
import java.util.Arrays;

class Solution {
    public static boolean findResult(int[] nums, int mid, int threshold) {
        int summ = 0;
        for (int x : nums) {
            summ += Math.ceil(x / (double) mid);
        }
        return summ <= threshold;
    }

    public int smallestDivisor(int[] nums, int threshold) {
        int low = 1;
        int high = Arrays.stream(nums).max().getAsInt() + 1; // Add 1 to include the maximum value
        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (mid == 0) {
                return 1; // Handle the case where mid is 0
            }
            if (findResult(nums, mid, threshold)) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return low;
    }
}

File: Step 04 Binary Search/Step 4.3 Find Answers by BS in Search Space/Find the Smallest Divisor Given a Threshold/Find the Smallest Divisor Given a Threshold.cpp
// class Solution {
// public:
//     int smallestDivisor(vector<int>& nums, int threshold) {
//         int minn = INT_MAX;
//         for(int i=*max_element(nums.begin(),end(nums));i>0;--i){
//             int sum = 0;
//             for(int j=0;j<nums.size();++j){
//                 sum += ceil(nums[j]/(float)i);
//             }
//             if (sum<=threshold)
//                 minn = min(minn,i);
//         }
//         return minn;
//     }
// };

class Solution {
public:
    int smallestDivisor(vector<int>& nums, int threshold) {
        int i=1, j=INT_MAX;
        while(i<j){
            int mid=i+(j-i)/2;
            int cnt=0;
            for(auto it: nums){
                cnt+=(it/mid)+(it%mid!=0);
            }
            //cout<<cnt<<endl;
            if(cnt>threshold){
                i=mid+1;
            }
            else{
                j=mid;
            }
        }
        return i;
    }
};